src/ompl/base/samplers/InformedStateSampler.h

```diff
@@ -0,0 +1,109 @@
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* Authors: Jonathan Gammell */
+
+#ifndef OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+#define OMPL_BASE_SAMPLERS_INFORMED_SAMPLER_
+
+// We inherit from StateSampler
+#include "ompl/base/StateSampler.h"
+// Deriving functions must be able to sample within a given cost
+#include "ompl/base/Cost.h"
+// We use a pointer to the problem definition to access problem and solution data.
+#include "ompl/base/ProblemDefinition.h"
+
+namespace ompl
+{
+    namespace base
+    {
+        OMPL_CLASS_FORWARD(InformedStateSampler);
+
+        /** \brief An abstract class for state samplers that use information
+        about the current solution to limit future search to a planning
+        subproblem that contains all possibly better solutions. */
+        class InformedStateSampler : public StateSampler
+        {
+        public:
+            /** \brief Construct a sampler that only generates states with a heuristic solution estimate that is less than the cost of the current solution. */
+            InformedStateSampler(const StateSpace* space, const ProblemDefinitionPtr probDefn, const Cost* bestCost);
+            virtual ~InformedStateSampler(void)
+            {
+            }
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are less than the current best cost (as defined by the pointer passed at construction). By default just calls sampleUniform(State*, Cost) with cost given by the member variable. */
+            virtual void sampleUniform(State* statePtr);
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are less than the provided cost. */
+            virtual void sampleUniform(State* statePtr, const Cost& maxCost) = 0;
+
+            /** \brief Sample uniformly in the subset of the state space whose heuristic solution estimates are between the provided costs. */
+            virtual void sampleUniform(State* statePtr, const Cost& minCost, const Cost& maxCost) = 0;
```

Comment:
Based on the existance of this function & the casting operations for samplers above, this makes me think samplers and informed samplers need to be treated differently, do they not? Basically, if a planner needs to cast a sampler to an InformedSampler, then we should probably have a separate base class for informed samplers.